pkg/serving/v1/apply.go

```diff
@@ -0,0 +1,268 @@
+// improved for a better merge strategy.
+func (cl *knServingClient) patch(modifiedService *servingv1.Service, currentService *servingv1.Service, uOriginalService []byte) (bool, error) {
+	uModifiedService, err := getModifiedConfiguration(modifiedService, true)
+	if err != nil {
+		return false, err
+	}
+	hasChanged, err := cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	for i := 1; i <= 5 && apierrors.IsConflict(err); i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		currentService, err = cl.GetService(currentService.Name)
+		if err != nil {
+			return false, err
+		}
+		hasChanged, err = cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	}
+	return hasChanged, err
+}
+
+func (cl *knServingClient) patchSimple(currentService *servingv1.Service, uModifiedService []byte, uOriginalService []byte) (bool, error) {
+	// Serialize the current configuration of the object from the server.
+	uCurrentService, err := encodeService(currentService)
+	if err != nil {
+		return false, err
+	}
+
+	patch, err := jsonmergepatch.CreateThreeWayJSONMergePatch(uOriginalService, uModifiedService, uCurrentService)
+	if err != nil {
+		return false, err
+	}
+
+	if string(patch) == "{}" {
+		return false, nil
+	}
+
+	// Check if the generation has been counted up, only then the backend detected a change
+	savedService, err := cl.patchService(currentService.Name, types.MergePatchType, patch)
+	if err != nil {
+		return false, err
+	}
+	return savedService.Generation != savedService.Status.ObservedGeneration, nil
+}
+
+// patchService patches the given service
+func (cl *knServingClient) patchService(name string, patchType types.PatchType, patch []byte) (*servingv1.Service, error) {
+	service, err := cl.client.Services(cl.namespace).Patch(context.TODO(), name, patchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return nil, err
+	}
+	err = updateServingGvk(service)
+
+	return service, err
+}
+
+func getOriginalConfiguration(service *servingv1.Service) []byte {
+	annots := service.Annotations
+	if annots == nil {
+		return nil
+	}
+	original, ok := annots[v1.LastAppliedConfigAnnotation]
+	if !ok {
+		return nil
+	}
+	return []byte(original)
+}
+
+func getModifiedConfiguration(service *servingv1.Service, annotate bool) ([]byte, error) {
+
+	// First serialize the object without the annotation to prevent recursion,
+	// then add that serialization to it as the annotation and serialize it again.
+	var uModifiedService []byte
+
+	// Otherwise, use the server side version of the object.
+	// Get the current annotations from the object.
+	annots := service.Annotations
+	if annots == nil {
+		annots = map[string]string{}
+	}
+
+	original := annots[v1.LastAppliedConfigAnnotation]
+	delete(annots, v1.LastAppliedConfigAnnotation)
+	service.Annotations = annots
+
+	uModifiedService, err := encodeService(service)
+	if err != nil {
+		return nil, err
+	}
+
+	if annotate {
```

Comment:
will there be a case when we'd want to control whether to annotate the service?

Reply:
Yes, there is. IIUR when you do the diff, then you don't want to have that annotation present. I would have to look a bit closer again, but there is a good reason. I also more or less took over the way how kubectl does the client side apply.